ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java

```diff
@@ -415,4 +424,64 @@ List<ChunkInfo> getChunksToStitch(CompleteMultipartUpload completeMultipartUploa
       return chunkInfos;
     }
   }
+
+  /**
+   * Check the list size and part number before processing request
+   * 1. Disallow duplicate part numbers
+   * 2. Disallow duplicate etags
+   * 3. Check for list size 10000
+   * 4. Check for part numbers integer 1-10000
+   * @param parts sorted parts list
+   * @return the bad request error
+   */
+  private static void validateParts(List<Part> parts) throws RestServiceException {
+    if (parts == null || parts.isEmpty()) {
+      throw new RestServiceException(S3Constants.ERR_EMPTY_REQUEST_BODY, RestServiceErrorCode.BadRequest);
+    }
+
+    if (parts.size() > S3Constants.MAX_LIST_SIZE) {
+      String error = S3Constants.ERR_PART_LIST_TOO_LONG;
+      throw new RestServiceException(error, RestServiceErrorCode.BadRequest);
+    }
+
+    Set<Integer> partNumbers = new HashSet<>();
+    Set<String> etags = new HashSet<>();
+
+    for (Part part : parts) {
+      int partNumber = part.getPartNumber();
+      String etag = part.geteTag();
+
+      if (partNumber < S3Constants.MIN_PART_NUM || partNumber > S3Constants.MAX_PART_NUM) {
+        String error = String.format(S3Constants.ERR_INVALID_PART_NUMBER, partNumber, S3Constants.MIN_PART_NUM,
+            S3Constants.MAX_PART_NUM);
+        throw new RestServiceException(error, RestServiceErrorCode.BadRequest);
+      }
+
+      if (!partNumbers.add(partNumber)) {
+        String error = String.format(S3Constants.ERR_DUPLICATE_PART_NUMBER, partNumber);
+        throw new RestServiceException(error, RestServiceErrorCode.BadRequest);
+      }
+
+      if (!etags.add(etag)) {
+        String error = String.format(S3Constants.ERR_DUPLICATE_ETAG, etag);
+        throw new RestServiceException(error, RestServiceErrorCode.BadRequest);
+      }
+    }
+  }
+
+  /**
+   * Check the parts array size is not empty or exceeds the limit
+   * @param parts the part array from Complete Multipart upload request
+   * @throws RestServiceException
+   */
+  private static void validatePartsSize(Part[] parts) throws RestServiceException {
```

Comment:
Remove this and just call validateParts first thing in the try-block. Make sure that has all the validations. This is what I am trying to get at.